File: Tests/parallel_switch.c
#include "acc_testsuite.h"

int test(){
    int err = 0;
    srand(time(NULL));
    int * a = (int *)malloc(n * sizeof(int));
    real_t * b = (real_t *)malloc(n * sizeof(real_t));
    real_t * b_host = (real_t *)malloc(n * sizeof(real_t));
    real_t * c = (real_t *)malloc(n * sizeof(real_t));


    for (int x = 0; x < n; ++x){
        a[x] = floor(rand() / (real_t)(RAND_MAX / 10));
        b[x] = rand() / (real_t)(RAND_MAX / 10);
        b_host[x] = b[x];
        c[x] = 0.0;
    }
    #pragma acc enter data copyin(a[0:n], b[0:n], c[0:n])
    #pragma acc parallel present(a[0:n], b[0:n], c[0:n])
    {
        #pragma acc loop
        for (int x = 0; x < n; ++x){
            switch(a[x]){
                case 0:
                    c[x] = b[x] * b[x];
                    break;
                case 1:
                    c[x] = b[x] / b[x];
                    break;
                case 2:
                    b[x] = b[x] / 2;
                default:
                    c[x] = a[x] + b[x];


            }
        }
    }
    #pragma acc exit data delete(a[0:n], b[0:n]) copyout(c[0:n])

    real_t tempc = 0.0;
    for (int x = 0; x < n; ++x){
        switch(a[x]){
            case 0:
                tempc = b_host[x] * b_host[x];
                break;
            case 1:
                tempc = b_host[x] / b_host[x];
                break;
            case 2:
                b_host[x] = b_host[x] / 2;
            default:
                tempc = a[x] + b_host[x];
        }
        if (fabs(c[x] - tempc) > PRECISION){
            err = 1;
        }
    }

    free(a);
    return err;
}


int main()
{
  int i;                        /* Loop index */
  int result;           /* return value of the program */
  int failed=0;                 /* Number of failed tests */
  int success=0;                /* number of succeeded tests */
  static FILE * logFile;        /* pointer onto the logfile */
  static const char * logFileName = "OpenACC_testsuite.log";        /* name of the logfile */


  /* Open a new Logfile or overwrite the existing one. */
  logFile = fopen(logFileName,"w+");

  printf("######## OpenACC Validation Suite V %s #####\n", ACCTS_VERSION );
  printf("## Repetitions: %3d                       ####\n",REPETITIONS);
  printf("## Array Size : %.2f MB                 ####\n",ARRAYSIZE * ARRAYSIZE/1e6);
  printf("##############################################\n");
  printf("Testing parallel_switch\n\n");

  fprintf(logFile,"######## OpenACC Validation Suite V %s #####\n", ACCTS_VERSION );
  fprintf(logFile,"## Repetitions: %3d                       ####\n",REPETITIONS);
  fprintf(logFile,"## Array Size : %.2f MB                 ####\n",ARRAYSIZE * ARRAYSIZE/1e6);
  fprintf(logFile,"##############################################\n");
  fprintf(logFile,"Testing parallel_switch\n\n");

  for ( i = 0; i < REPETITIONS; i++ ) {
    fprintf (logFile, "\n\n%d. run of parallel_switch out of %d\n\n",i+1,REPETITIONS);
    if (test() == 0) {
      fprintf(logFile,"Test successful.\n");
      success++;
    } else {
      fprintf(logFile,"Error: Test failed.\n");
      printf("Error: Test failed.\n");
      failed++;
    }
  }

  if(failed==0) {
    fprintf(logFile,"\nDirective worked without errors.\n");
    printf("Directive worked without errors.\n");
    result=0;
  } else {
    fprintf(logFile,"\nDirective failed the test %i times out of %i. %i were successful\n",failed,REPETITIONS,success);
    printf("Directive failed the test %i times out of %i.\n%i test(s) were successful\n",failed,REPETITIONS,success);
    result = (int) (((double) failed / (double) REPETITIONS ) * 100 );
  }
  printf ("Result: %i\n", result);
  return result;
}

File: Tests/parallel_switch.f90
      INTEGER FUNCTION test()
        IMPLICIT NONE
        INCLUDE "acc_testsuite.fh"
        INTEGER :: x !Iterators
        INTEGER,DIMENSION(LOOPCOUNT):: a
        REAL(8),DIMENSION(LOOPCOUNT):: b, c, randoms!Data
        INTEGER :: errors = 0

        !Initilization
        CALL RANDOM_SEED()
        CALL RANDOM_NUMBER(b)
        c = 0
        CALL RANDOM_NUMBER(randoms)
        a = floor(randoms * 3)

        !$acc enter data copyin(a(1:LOOPCOUNT), b(1:LOOPCOUNT), &
            !$acc c(1:LOOPCOUNT))
        !$acc parallel present(a(1:LOOPCOUNT), b(1:LOOPCOUNT), &
            !$acc c(1:LOOPCOUNT))
          !$acc loop
          DO x = 1, LOOPCOUNT
            SELECT CASE (a(x))
              CASE (0)
                c(x) = b(x) * b(x)
              CASE (1)
                c(x) = b(x) / b(x)
              CASE (2)
                c(x) = b(x) / 2
            END SELECT
          END DO
        !$acc end parallel
        !$acc exit data delete(a(1:LOOPCOUNT), b(1:LOOPCOUNT)) &
            !$acc copyout(c(1:LOOPCOUNT))

        DO x = 1, LOOPCOUNT
          SELECT CASE (a(x))
            CASE (0)
              IF (abs(c(x) - (b(x) * b(x))) .gt. PRECISION) THEN
                errors = errors + 1
              END IF
            CASE (1)
              IF (abs(c(x) - (b(x) / b(x))) .gt. PRECISION) THEN
                errors = errors + 1
              END IF
            CASE (2)
              IF (abs(c(x) - (b(x) / 2)) .gt. PRECISION) THEN
                errors = errors + 1
              END IF
          END SELECT
        END DO

        test = errors
      END


      PROGRAM test_kernels_async_main
      IMPLICIT NONE
      INTEGER :: failed, success !Number of failed/succeeded tests
      INTEGER :: num_tests,crosschecked, crossfailed, j
      INTEGER :: temp,temp1
      INCLUDE "acc_testsuite.fh"
      INTEGER test


      CHARACTER*50:: logfilename !Pointer to logfile
      INTEGER :: result

      num_tests = 0
      crosschecked = 0
      crossfailed = 0
      result = 1
      failed = 0

      !Open a new logfile or overwrite the existing one.
      logfilename = "OpenACC_testsuite.log"
!      WRITE (*,*) "Enter logFilename:"
!      READ  (*,*) logfilename

      OPEN (1, FILE = logfilename)

      WRITE (*,*) "######## OpenACC Validation Suite V 2.5 ######"
      WRITE (*,*) "## Repetitions:", N
      WRITE (*,*) "## Loop Count :", LOOPCOUNT
      WRITE (*,*) "##############################################"
      WRITE (*,*)

      WRITE (*,*) "--------------------------------------------------"
      WRITE (*,*) "Test of parallel_switch"
      WRITE (*,*) "--------------------------------------------------"

      crossfailed=0
      result=1
      WRITE (1,*) "--------------------------------------------------"
      WRITE (1,*) "Test of parallel_switch"
      WRITE (1,*) "--------------------------------------------------"
      WRITE (1,*)
      WRITE (1,*) "(Crosstests should fail)"
      WRITE (1,*)

      DO j = 1, N
        temp =  test()
        IF (temp .EQ. 0) THEN
          WRITE (1,*)  j, ". test successfull."
          success = success + 1
        ELSE
          WRITE (1,*) "Error: ",j, ". test failed."
          failed = failed + 1
        ENDIF
      END DO


      IF (failed .EQ. 0) THEN
        WRITE (1,*) "Directive worked without errors."
        WRITE (*,*) "Directive worked without errors."
        result = 0
        WRITE (*,*) "Result:",result
      ELSE
        WRITE (1,*) "Directive failed the test ", failed, " times."
        WRITE (*,*) "Directive failed the test ", failed, " times."
        result = failed * 100 / N
        WRITE (*,*) "Result:",result
      ENDIF
      CALL EXIT (result)
      END PROGRAM


